T5gemma2

bzhangGo opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Add support for T5Gemma2 with multi-modal and long-context capability.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker @Cyrilvallez
- vision models: @yonigozlan @molbap
- audio models: @eustlb @ebezzam @vasqu
- multimodal models: @zucchini-nlp
- graph models: @clefourrier

Library:

- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- continuous batching: @remi-or @ArthurZucker @McPatate
- pipelines: @Rocketknight1
- tokenizers: @ArthurZucker and @itazap
- trainer: @SunMarc
- attention: @vasqu @ArthurZucker @CyrilVallez
- model loading (from pretrained, etc): @CyrilVallez
- distributed: @3outeille @ArthurZucker
- CIs: @ydshieh

Integrations:

- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization: @SunMarc @MekkCyber
- kernels: @MekkCyber @drbh
- peft: @BenjaminBossan @githubnemo

Devices/Backends:

- AMD ROCm: @ivarflakstad
- Intel XPU: @IlyasMoutawwakil
- Ascend NPU: @ivarflakstad 

Documentation: @stevhliu

Research projects are not maintained and should be taken as is.

 -->

bzhangGo committed:
Fix small bug in T5Gemma 1 in __init__

bzhangGo committed:
Add t5gemma2 model & configurations.

bzhangGo committed:
Add auto support

bzhangGo committed:
Add test case.

bzhangGo committed:
Add doctree.

bzhangGo committed:
Update positional embeddings to match latest update.

bzhangGo committed:
Style fix & add use of final_logit_softcapping for attributes check.

bzhangGo committed:
Update tests and embedding design.

bzhangGo committed:
Add t5gemma2 to image-text-to-text category.

bzhangGo committed:
Add T5Gemma2 doc.

bzhangGo committed:
remove unused imports.

Rocketknight1 commented:
Hey! Is this a pre-release model? I don't see the checkpoints like `google/t5gemma-2-270m` anywhere

Rocketknight1 commented:
cc @arthurzucker in that case!

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+)
+
+
+_CHECKPOINT_FOR_DOC = "google/t5gemma-2-270m-270m"
```
Shouldnt be needed anymore

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+        vocab_size (`int`, *optional*, defaults to 262144):
+            Vocabulary size of the T5Gemma2 model (the same as Gemma 3).
+        kwargs (additional keyword arguments, optional, *optional*):
+            Will be passed to the PreTrainedConfig base class.
```
```suggestion
```

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+        seq_len: Optional[int] = None,
+        layer_type: Optional[str] = None,
+    ) -> tuple["torch.Tensor", float]:
+        return super().compute_default_rope_parameters(config, device, seq_len, layer_type)
```
```suggestion
class T5Gemma2RotaryEmbedding(Gemma3RotaryEmbedding):
    pass
```
We do not modify anything, so we can just pass and directly inherit everything, no?

vasqu replied:
Ah, thx for clarifying that's annoying. Let's keep it this way then

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+
+class T5Gemma2SelfAttention(Gemma3Attention):
+    def __init__(self, config: T5Gemma2ModuleConfig, layer_idx: int):
+        super().__init__(config, layer_idx)
```
```suggestion
class T5Gemma2SelfAttention(Gemma3Attention):
    pass
```
Same here

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+            image_features, image_mask = self.pixel2feature_preprocessor_fn(pixel_values, input_ids, inputs_embeds)
+
+            image_features = image_features.to(inputs_embeds.device, inputs_embeds.dtype)
+            inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_features)
```
Do we have to do this? Can it not be handled in for conditional model? We really shouldn't be passing functions around tbh

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+        hidden_states = self.dropout(hidden_states)
+
+        for layer_module in self.layers[: self.config.num_hidden_layers]:
+            assert isinstance(layer_module, T5Gemma2EncoderLayer)
```
no asserts

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+        """Convert pixel images to image features and obtain placeholder mask."""
+        image_features = self._get_image_features(pixel_values)
+        image_mask = self._get_placeholder_mask(input_ids, inputs_embeds=inputs_embeds, image_features=image_features)
+        return image_features, image_mask
```
I'd be pro applying this here instead of passing things around and then doing it in the encoder if possible

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+            past_key_values=decoder_outputs.past_key_values,
+            decoder_hidden_states=decoder_outputs.hidden_states
+            if kwargs.get("output_hidden_states", False)
+            else (decoder_outputs.last_hidden_state,),
```
```suggestion
            decoder_hidden_states=decoder_outputs.hidden_states,
```

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+        """
+        # encoder
+        if encoder_outputs is None:
+            encoder_outputs = self.encoder(
```
You can pass return_dict=True for both encoder and decoder just to be safe

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+
+        # setup encoder and decoder
+        self.encoder = T5Gemma2Encoder(config.encoder, config.eoi_token_index, self.pixel2feature_preprocessor)
+        self.decoder = T5Gemma2Decoder(config.decoder, config.eoi_token_index)
```
Should be wrapped under language model so that we treat as a complete model here

vasqu reviewed (commented):
Just an initial review from my side

3 main issues I would have are
- We definitely should be able to have a SWA for bidirectional masks directly in the utils. 
- The process fn to process pixel values should not be passed around imo, it should be handled a level above
- We should have one level for the encoder, decoder models to be wrapped under language model

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
````diff
@@ -0,0 +1,1445 @@
+    >>> from transformers import T5Gemma2Config, T5Gemma2Model
+    >>> t5gemma2_config = T5Gemma2Config.from_pretrained("google/t5gemma-270m-270m")
+    >>> model = T5Gemma2Model(t5gemma2_config)
+    ```
````
ok.

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+            encoder = T5Gemma2ModuleConfig()
+            logger.info("encoder is None, using default T5Gemma2ModuleConfig encoder config.")
+        else:
+            assert isinstance(encoder, T5Gemma2ModuleConfig), f"{type(encoder)} is not supported."
```
sure, will replace all assert with an error.

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+
+        for special_token_key in ["bos_token_id", "pad_token_id", "eos_token_id"]:
+            if special_token_key not in kwargs:
+                kwargs[special_token_key] = getattr(decoder, special_token_key)
```
Yeah, these special tokens are already part of the encoder/decoder config (see T5Gemma2ModuleConfig), and here we adopted the setup from decoder by default. I just want to avoid add these duplicate defintions.

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+            raise ValueError(
+                "Imbalanced encoder-decoder is not supported in T5Gemma2: "
+                f"encoder ({encoder_hidden_states.shape[-1]}) vs decoder ({hidden_states.shape[-1]})."
+            )
```
ok.

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+            Pixel values to be processed by the image encoder to extract image features.
+        """
+        # Part of Gemma 3 processor output but not used by t5gemma 2.
+        del token_type_ids
```
So, in T5Gemma 2, we don't need the token_type_ids as we perform bidirectional attention directly for all encoder inputs. Howerver, we re-used the preprocessor defined in Gemma 3, which produces token_type_ids by default.

So to simplify the pipeline, we just added token_type_ids as an argument and delete it in the function (just a mark that this is not used).

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+
+        if position_ids is None:
+            position_ids = cache_position.unsqueeze(0)
+
```
yeah

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+            raise ValueError(
+                "Imbalanced encoder-decoder vocabulary size is not supported in T5Gemma2: "
+                f"encoder ({config.encoder.vocab_size}) vs decoder ({config.decoder.vocab_size})."
+            )
```
done

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+    _dynamic_tied_weights_keys = [
+        "model.encoder.embed_tokens.eoi_embedding",
+        "model.decoder.embed_tokens.eoi_embedding",
+    ]
```
yeah, otherwise, saving in safetensor will complain.

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+    _pp_plan = {"lm_head.out_proj": (["hidden_states"], ["logits"])}
+
+    def __init__(self, config: T5Gemma2Config):
+        config.is_encoder_decoder = True
```
done.

bzhangGo committed:
minor update following comments.

bzhangGo committed:
minor style fixes.

bzhangGo committed:
fix config.

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+        right_mask = (dist < 0) & (-dist < right_window_size)
+        return left_mask | right_mask
+
+    return inner_mask
```
Gotcha, that's fine! I will work on creating support for this. Just FYI, this will have slight perf regressions as it uses vmap + it's only supported for torch 2.6 and up

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1381 @@
+        self.vocab_size = config.vocab_size
+
+        # preprocessor for raw images pixel values: injected from outside.
+        self.pixel2feature_preprocessor_fn = pixel2feature_preprocessor_fn
```
We already discussed this a bit and I like your idea with the encoder directly having this better.

So, we move all this into the encoder code explicitly and for the level above xxxModel, we call our encoder with the implemented functions instead. Parent -> child is nicer than the current child -> parent dependency (if that makes sense).

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+            position_ids = cache_position.unsqueeze(0)
+
+        if attention_mask is None:
+            attention_mask = make_default_2d_attention_mask(input_ids, inputs_embeds, self.config.pad_token_id)
```
Can you check if we really need this? It uses some old APIs we don't really want to use and in most cases we could remove this without issues in some older models like T5 (WIP but soon merged)

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1381 @@
+    _dynamic_tied_weights_keys = [
+        "encoder.embed_tokens.eoi_embedding",
+        "decoder.embed_tokens.eoi_embedding",
+    ]
```
Arthur is working on making this easier just fyi, see #41580

vasqu reviewed (commented):
The key change should be how we handle caches and the encoder with vision preprocessing (get image features etc). 

Regarding generation related issue, for now we should override the respective functions in the generation mixin from which we inherit. I agree that we ideally have proper logic in our code, but I want to postpone this for now and fix this properly in the future; small overrides are fine and we always encounter these here and there already.

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1381 @@
+        vocab_size (`int`, *optional*, defaults to 262144):
+            Vocabulary size of the T5Gemma2 model (the same as Gemma 3).
+        kwargs (additional keyword arguments, optional, *optional*):
+            Will be passed to the PreTrainedConfig base class.
```
sure!

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+
+    def forward(self, input_ids: torch.Tensor):
+        input_embeddings = super().forward(input_ids) * self.embed_scale.to(self.weight.dtype)
+        input_embeddings[input_ids == self.eoi_token_index] = self.eoi_embedding.to(input_embeddings.dtype)
```
ok! done!

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1381 @@
+        self.patches_per_image = int(config.vision_config.image_size // config.vision_config.patch_size)
+        self.tokens_per_side = int(config.mm_tokens_per_image**0.5)
+        self.kernel_size = self.patches_per_image // self.tokens_per_side
+        self.avg_pool = nn.AvgPool2d(kernel_size=self.kernel_size, stride=self.kernel_size)
```
yeah, let's reuse the shared part.

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1445 @@
+            position_ids = cache_position.unsqueeze(0)
+
+        if attention_mask is None and past_key_values is None:
+            attention_mask = make_default_2d_attention_mask(input_ids, inputs_embeds, self.config.pad_token_id)
```
seems fine to remove it.

bzhangGo committed:
Update T5Gemma2 following Anton's comments:
1. Override _prepare_cache_for_generation to take care of cross-attention cache.
2. Move vision preprocessing from main model to encoder.
3. Clean and fix bugs in modular model.

bzhangGo committed:
Add T5Gemma2VisionConfig.

bzhangGo committed:
Minor updates.

bzhangGo committed:
fix style

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1500 @@
+        if vision_config is not None:
+            self.vision_encoder = T5Gemma2VisionEncoder(vision_config)
+        else:
+            self.vision_encoder = None
```
Do we have cases where this is `None`? We should avoid if else branches like these and just clearly define them

Also, might not need to be a separate module then

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1500 @@
+            Pixel values to be processed by the image encoder to extract image features.
+        """
+        # Unused for processor compatibility kept in signature.
+        del token_type_ids
```
```suggestion
```
Just move the comment to the signature instead

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1500 @@
+        )
+
+
+class T5Gemma2Decoder(T5Gemma2Encoder):
```
Let's move the inheritance between modules within one file --> we only like to keep inheritances between different models - was also recently changed in the first t5gemma. This will have you probably create more duplicate code but it is what it is.

vasqu commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1500 @@
+
+        # cross-attention does not use sliding window
+        del cross_attn_config.sliding_window
+        del cross_attn_config.layer_types
```
Same for what I commented in the forward init version: Can you try initializing without a config? I'm pretty sure that should work but you can prove me wrong :D

bzhangGo replied:
will still need it in the case of staticcache

vasqu commented on tests/models/t5gemma2/test_modeling_t5gemma2.py:
```diff
@@ -0,0 +1,1130 @@
+    # used in `test_torch_compile_for_training`
+    _torch_compile_train_cls = T5Gemma2ForConditionalGeneration if is_torch_available() else None
+    # `T5Gemma2` will give warning or raise error if it is not `eager` during training.
+    _torch_compile_train_attn_implementation = "eager"
```
Can you check if this is still the case?

vasqu reviewed (commented):
Thank you, looking already pretty good. There are no big issues tbh, it's just few smaller issues for standards

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1500 @@
+
+    def __init__(
+        self,
+        siglip_config: Optional[Union[SiglipVisionConfig, dict[str, Any]]] = None,
```
then let the visionconfig here inherient from siglipconfig.

bzhangGo commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1500 @@
+        "decoder.layers": (["hidden_states", "attention_mask"], ["hidden_states"]),
+        "decoder.norm": (["hidden_states"], ["hidden_states"]),
+    }
+
```
done

bzhangGo commented on tests/models/t5gemma2/test_modeling_t5gemma2.py:
```diff
@@ -0,0 +1,1130 @@
+    _torch_compile_train_attn_implementation = "eager"
+
+    # won't fix
+    test_torchscript = False
```
also removed.

bzhangGo committed:
re-structure vision encoder and minor fixes.

bzhangGo committed:
fix parameter tying.

vasqu commented on tests/models/t5gemma2/test_modeling_t5gemma2.py:
```diff
@@ -0,0 +1,1133 @@
+    # Based on tests.test_modeling_common.ModelTesterMixin.test_flex_attention_with_grads
+    # Update hidden size for encoder and decoder
+    @require_torch_accelerator
+    def test_flex_attention_with_grads(self):
```
Hmm, we changed the test to iterate through all subconfigs so I'd hope we could avoid overriding this. Same for the tests above - a lot of overrides. Would be nice if we could avoid some

vasqu reviewed (approved):
Thanks a lot for iterating! I left some last comments which are mostly things to shorten / simplify things. 

If we could remove some test overrides, that would be awesome. Atm, it's a lot which is unideal but also not the world.

bzhangGo commented on tests/models/t5gemma2/test_modeling_t5gemma2.py:
```diff
@@ -0,0 +1,1133 @@
+
+    # MP works but offload doesn't work when the SigLIP MultiheadAttention is offloaded
+    # TODO: One potential solution would be to add to set preload_module_classes = ["SiglipMultiheadAttentionPoolingHead"]
+    # in the dispatch_model function
```
I didn't add it to _no_split_modules. This comments was copied from Gemma 3.

bzhangGo committed:
remove several unnecessary codes and fix small bugs.

bzhangGo committed:
update and fix init.

ArthurZucker commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1356 @@
+                [self_attn_mask_mapping[attn_type], cross_attn_mask_mapping["full_attention"]], dim=-1
+            )
+            for attn_type in self_attn_mask_mapping
+        }
```
there are only 2 attention type so we can just be explicit here

ArthurZucker commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1356 @@
+        # Decoder input and output embeddings are tied.
+        if self.config.tie_word_embeddings:
+            self.decoder.embed_tokens.weight = self.encoder.embed_tokens.weight
+            self.decoder.embed_tokens.eoi_embedding = self.encoder.embed_tokens.eoi_embedding
```
should now work OOB

ArthurZucker commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1356 @@
+        "lm_head.out_proj.weight": "model.encoder.embed_tokens.weight",
+        "model.decoder.embed_tokens.weight": "model.encoder.embed_tokens.weight",
+        "model.decoder.embed_tokens.eoi_embedding": "model.encoder.embed_tokens.eoi_embedding",
+    }
```
```suggestion
    _tied_weights_keys = {
        "lm_head.out_proj.weight": "model.encoder.embed_tokens.weight",
    }
```

ArthurZucker replied:
child is responsible for himself :)

ArthurZucker commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1356 @@
+        # Decoder input and output embeddings are tied.
+        if self.config.tie_word_embeddings:
+            self.lm_head.out_proj.weight = self.model.encoder.embed_tokens.weight
+
```
```suggestion
```

ArthurZucker commented on src/transformers/models/t5gemma2/modular_t5gemma2.py:
```diff
@@ -0,0 +1,1356 @@
+    _tied_weights_keys = {
+        "model.decoder.embed_tokens.weight": "model.encoder.embed_tokens.weight",
+        "model.decoder.embed_tokens.eoi_embedding": "model.encoder.embed_tokens.eoi_embedding",
+    }
```
```suggestion
```

ArthurZucker commented on src/transformers/models/t5gemma2/modeling_t5gemma2.py:
```diff
@@ -0,0 +1,1572 @@
+        return f"{tuple(self.weight.shape)}, eps={self.eps}"
+
+
+class T5Gemma2MLP(nn.Module):
```
inherit from T5GemmaMLP

ArthurZucker commented on src/transformers/models/t5gemma2/modeling_t5gemma2.py:
```diff
@@ -0,0 +1,1572 @@
+        return hidden_states
+
+
+class T5Gemma2DecoderLayer(T5Gemma2EncoderLayer):
```
This is not really expected we should disantangle them!

ArthurZucker reviewed (commented):
Very nice work!

bzhangGo committed:
Update weight tying and other minor changes.

bzhangGo committed:
Skip `tie_word_embeddings` in config attributes check in T5Gemma2.

ArthurZucker commented on tests/models/t5gemma2/test_modeling_t5gemma2.py:
I don't see integration tests, would be better to have them!

ArthurZucker reviewed (approved):
Thanks for adressing most comments!

bzhangGo committed:
minor fix.

bzhangGo committed:
fix the inherence of t5gemma2decoderlayer

bzhangGo committed:
sync to head.

vasqu committed:
Merge branch 'main' into t5gemma2

vasqu committed:
update decorator usage

vasqu committed:
disable FA and Flex due to merged module behavior

vasqu committed:
style
